src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/themes/AceThemes.java

```diff
@@ -55,22 +55,48 @@ public AceThemes(ThemeServerOperations themeServerOperations,
       events_ = events;
       prefs_ = prefs;
       themes_ = new HashMap<>();
+      docRootUrl_ = null;
 
       prefs.get().theme().bind(theme -> applyTheme(theme));
    }
    
    private void applyTheme(Document document, final AceTheme theme)
    {
-      Element oldStyleEl = document.getElementById(linkId_);
+      // Build a relative path to avoid having to register 80000 theme URI handlers on the server.
+      int pathUpCount = 0;
+      String currentUrl = document.getURL();
+      if (null == docRootUrl_)
```

Comment:
Ah, the URL is like `"https://127.0.0.1:8787"`, if it were running behind a reverse proxy with a path set up, it might look like `"https://serverhost/rstudio"`. The first call to apply theme should always be for the root document (i.e. not a nested document), and the part that I have to negate is the difference between the URL of the root document and the URL of the nested document. For example, the URL of the help pane showing `xml2::xml_name` would be `"https://serverhost/rstudio/help/library/xml/xml_name"`, and I want to negate the `help/library/xml/xml_name` part of it. To get that part, I need to know the concrete URL of the root document, which is why I save it the first time around.

Reply:
Got it, thanks! I think we should avoid the order dependency -- it's true that `applyTheme` is always called by the root document first today but there's a possibility that won't hold true in the future. I'd suggest either (a) initializing the URL in the constructor, or (b) using `GWT.getHostPageBaseURL()`, which would allow you to do this totally statelessly.